VitalClient/src/main/java/io/tryvital/client/userConnectedSources.kt

```diff
@@ -1,23 +1,62 @@
 package io.tryvital.client
 
-import io.tryvital.client.services.data.ManualProviderRequest
-import io.tryvital.client.services.data.ManualProviderSlug
+import io.tryvital.client.services.data.*
 import retrofit2.HttpException
 
+fun VitalClient.hasUserConnectedTo(provider: ProviderSlug): Boolean {
+    return this.sharedPreferences.getBoolean(
+        VitalClientPrefKeys.userHasConnectedTo(provider),
+        false
+    )
+}
+
 // TODO: VIT-2924 Move userId management to VitalClient
-suspend fun VitalClient.createConnectedSource(provider: ManualProviderSlug, userId: String) {
+suspend fun VitalClient.createConnectedSourceIfNotExist(provider: ManualProviderSlug, userId: String) {
+    val slug = provider.toProviderSlug()
+    if (hasUserConnectedTo(slug)) {
+        // Local Hit: The client has witnessed a valid connected source for this provider before.
+        return
+    }
+
+    // Local Miss: First try to query the user's current set of connected sources.
+    val sources = userConnectedSources(userId = userId)
+    if (sources.any { it.slug == slug }) {
+        // Remote Hit: The client has connected to this provider.
+        return
+    }
+
     try {
+        // Remote Miss: Try to create the manual connected source.
         linkService.manualProvider(
             provider = provider,
             request = ManualProviderRequest(userId = userId)
         )
     } catch (exception: HttpException) {
-        if (exception.code() == 409) {
-            // A 409 means that there's a already a connected source, so we can fail gracefully.
-            return
-        } else {
-            // Rethrow exception.
+        // A 409 means that there's a already a connected source, so we can fail gracefully.
+        if (exception.code() != 409) {
+            // Rethrow any other exception.
             throw exception
         }
```

Comment:
Fail by throwing an exception?

Reply:
409 Conflict means a connected source already exists, so we don’t need to fail the call.